### DISCUSSION TITLE
Network resilience with Polly

### DISCUSSION THREAD

**Discussion by henda79** [Azure/azure-mobile-apps#700] [General]
I was looking at handing network failures and adding resilience to my XF project using Polly.

I was using Polly in a MAUI project and was fairly straight forward as named `HttpClient`s can be added via the `IServiceCollection` and then later used.

However in the XF project we don't have this so was considering the best way to go about this. The only real solution I had come up with was to use a `DelegatingHandler` and add it to the pipeline with `PolicyHttpMessageHandler` from `Microsoft.Extensions.Http` since we can't get access the HttpClient.

What do other think about this ?

**Comment by adrianhall** [comment]
Perhaps I can provide an interface in the options to use; something like this:

``` csharp
public interface IHttpClientFactory
{
  HttpClient GetHttpClient(HttpMessageHandler rootHandler);
}
```

Internally, this will just do the obvious (`new HttpClient(rootHandler);`), but you'll be able to swap it out for anything you want.  Would that work for you?

**Reply by adrianhall** [reply]
The `PolicyHttpMessageHandler` is probably the right way to do it right now, though.

**Reply by adrianhall** [reply]
... and I'm sure you've seen this already: https://stackoverflow.com/questions/66553465/correct-way-to-retry-httpclient-requests-with-polly

**Reply by henda79** [reply]
Thanks Adrian, I've done quite a bit of research on using Polly over the last few days as its new to me, but most of the examples use the builder which is suited for MAUI etc.

With XF, I *think* the best way is to use the `PolicyHttpMessageHandler` and add it to the `HttpPipeline` which is exposed through the datasync client

What point should it be placed in the pipeline ? What order is it processed in ?

**Reply by adrianhall** [reply]
You should put the retry policy as the last item in the pipeline.

**Comment by henda79** [comment]
@adrianhall  I'm back to this now as the XF project is now MAUI, and was looking as to why we can't pass our HttpClient to the DataSync Client. It seems to make more sense as we can use the builder and extension methods for Polly / Simmy etc. I know this can be done with a DelegatingHandler but that seems a lot harder.

Also when it comes to testing I would think this way would easier too ?

**Reply by adrianhall** [reply]
I agree - take a look at the HttpClient stuff I’m writing for v8.  You’ll notice there is a IHttpClientProvider (which is basically IHttpClientFactory, but that one was taken) with a default implementation (imaginatively named DefaultHttpClientProvider).

I’m not happy with the API on this yet - it needs some real code writing around it, but should be enough to give you an idea of what I am thinking.

**Reply by henda79** [reply]
ok, that sounds great, even the "default" name :)

So for now I'll just stick with the DelegatingHandler to inject Polly.

I've been trying to keep up the v8 news but don't recall a release date, do you have anything in mind ?

**Reply by adrianhall** [reply]
Nothing concrete.  However, I'm aiming for the BUILD conference or thereabouts (which is really just to put a stake in the ground on a date rather than something I'm doing for BUILD)
